src/Dsll.hs

```diff
@@ -0,0 +1,397 @@
+
+rws :: [String] -- The list of reserved words in DSLL
+rws = ["constructor","operator","forvars","fortypes","predicate"]
+
+--Identifiers handling
+identifier :: Parser String
+identifier = (lexeme . try) (p >>= check)
+  where
+    p       = (:) <$> letterChar <*> many alphaNumChar
+    check x = if x `elem` rws
+                then fail $ "keyword " ++ show x ++ " cannot be an identifier"
+                else return x
+
+
+-- --------------------------------------- DSLL Parser -------------------------------------
+
+
+-- | 'DSLLParser' is the top-level parser function. The parser contains a list of functions
+--    that parse small parts of the language. When parsing a source program, these functions are invoked in a top-down manner.
+dsllParser :: Parser DSLLProg
+dsllParser = between scn eof dsllProg -- Parse all the statemnts between the spaces to the end of the input file
+
+-- |'dsllProg' parses the entire actual DSLL program which is a collection of constructors followed by a collection of
+--   operations followed by a collection of predicates
+dsllProg :: Parser DSLLProg
+dsllProg = do
+    listCd <- cdParser `sepEndBy` newline'
+    listOd <- odParser `sepEndBy` newline'
+    listPd <- pdParser `sepEndBy` newline'
+    return DSLLProg {cd = listCd, od =  listOd, pd =  listPd}
+
+
+typeParser :: Parser Type
+typeParser = do
+    i <- identifier
+    return i
+
+varParser :: Parser Var
+varParser = do
+    i <- identifier
+    return i
+
+typeVarParser :: Parser TypeVar
+typeVarParser = do
+    i <- identifier
+    return i
+
+propParser :: Parser Prop
+propParser = do
+    i <- identifier
+    return i
+
+
+tParser, tConstructorInvokerParser, typeVarParser' :: Parser T
+tParser = try tConstructorInvokerParser <|> typeVarParser'
+tConstructorInvokerParser = do
+    i         <- identifier
+    arguments <- parens ( argParser `sepBy1` comma)
+    return (TConstr (ConstructorInvoker {nameCons = i, argCons = arguments}))
+typeVarParser' = do
+    i <- typeVarParser
+    return (TTypeVar i)
+
+
+
+argParser, varParser', tParser'  :: Parser Arg
+argParser = try tParser' <|> varParser'
+varParser' = do
+     i <- varParser
+     return (AVar i)
+tParser' = do
+    t <- tParser
+    return (AT t)
+
+
+kParser, kTypeParser, tParser'' :: Parser K
+kParser = try kTypeParser <|> try tParser''
+kTypeParser = do
+     i <- typeParser
+     return (Ktype i)
+tParser'' = do
+    t <- tParser
+    return (KT t)
+
+cdParser, cd1, cd2 :: Parser Cd
+cdParser = try cd1 <|> cd2
+cd1= do
+    rword "constructor"
+    name <- identifier
+    colon
+    k' <- listOut (kParser `sepBy1` comma)
+    arrow
+    t' <- typeParser
+    return Cd {nameCd = name, inputCd =  k', outputCd =  t'}
+cd2 = do
+    rword "constructor"
+    name <- identifier
+    colon
+    t' <- typeParser
+    return Cd {nameCd = name, inputCd =  [], outputCd =  t'}
+
+odParser, od1, od2, od3, od4 :: Parser Od
```

Comment:
I wonder if there's some way to combine these cases (for cd, od, and pd)--having 0 or 1 of forvals or fortypes in any combination. (You've probably already thought of this :-))

Reply:
Sure I did, that's why I wrote the code need a major refactoring because it looks bad ;-))